lesson_23/headers/inplace_array.h

```diff
@@ -0,0 +1,324 @@
+#pragma once
+
+#include <variant>
+
+#include "debug_macroses.hpp"
+
+
+template <class T, int SmallBufferSize = 3>
+class inplace_array
+{
+public:
+    inplace_array();
+
+    inplace_array(const std::size_t size);
+
+    inplace_array(const inplace_array& other);
+
+    ~inplace_array() = default;
+
+public:
+    inplace_array& operator=(const inplace_array& other);
+
+    T& operator[](std::size_t index);
+
+    void set_size(const std::size_t new_size);
+
+    std::size_t get_size() const;
+
+    void clear();
+
+    void push_back(T element);
+
+    void pop_back();
+
+    T& back() const;
+
+    void reserve(const std::size_t reserved_space);
+
+    std::size_t get_capacity() const;
+
+    void shrink_to_fit();
+
+    bool operator==(const inplace_array& other);
+
+private:
+    bool small_buffer_used() const;
+
+    T* get_buffer_prt();
+
+    const T* get_buffer_prt() const;
+
+private:
+    struct small_buffer
+    {
+        T data[SmallBufferSize]{};
+    };
+
+    struct buffer
+    {
+        buffer() = default;
+
+        explicit buffer(const std::size_t size) : data{ new T[size] }
+        {
+        }
+
+        ~buffer()
+        {
+            delete[] data;
+            data = nullptr;
+        }
+
+        buffer(buffer&& other) noexcept
+        {
+            data = other.data;
+            other.data = nullptr;
+        }
+
+        buffer& operator=(buffer&& other) noexcept
+        {
+            if (this == &other)
+            {
+                return *this;
+            }
+            data = other.data;
+            other.data = nullptr;
+            return *this;
+        }
+
+        T* data{ nullptr };
+    };
```

Comment:
you don't need it.  you should create T bufer[SmallBufferSize] on stack, and use this SmallBufferSize as a capacity. If size will be bigger then capacity you should allocate memory on heap in the same way as dynamic array

Reply:
I guess I don't really get your idea. What do you mean by this _"create T bufer[SmallBufferSize] on stack"_.
AFAK there is no options for C++ class developer to control where memory for the members will be allocated for this class on the stack or on the heap. I can only control heap allocation for the members not stack, stack is up to class user.
I as a user of a class can decide it, but also there is no full control for a class members too (for some members class developer can decide to allocate memory on the heap)

**e.g.**
```
class A{}

class B
{
private:
   A a{};
}

class C
{
private:
   A* a = new A();
}

main()
{
    A a; //stack allocation
    A* a1 = new A(); //heap allocation

    B b; // A as member and B located on the stack
    B* b1 = new B();// A as member and B located on the heap
    //I can not control where A will be allocated;

    C c; // C on the stack, but B on the heap;
    C* c1 = new C(); // C and B on the heap;
    // only heap option for the B
}
```
So for this task I tried to implement sort of SBO (small buffer optimization) approach.
It was possible to use _union_ without additional class-wrappers
```
union buffer
{
    T bufer_[SmallBufferSize];
    T* buffer_heap_;
}
```
but in the modern C++ _std::variant_ with RAII since 17 standard is more safe and preferable.
I suppose if user will use inplace_array like this:
```
inplace_array* arr = new inplace_array();
```
there is no options to allocate inner buffer on the stack,
only in case like this:
```
inplace_array arr{};
```
and for memory saving purpose use _union_ or _std::variant_.

Where I am wrong?

Reply:
All that can be on stack, should be on stack. 
If we go further, we can forbid new operator for this class.
Generally unless your class contains pointers then the entire object is contained in one allocation

Reply:
Ok, these are general and obvious things. And it's not the answer on my question. What is wrong in my code, I don't see why it won't work properly.
If you use class like this
```
inplace_array arr{};
```
everything will be on stack until you exceed the limit of SmallBufferSize. Than buffer will be allocated on the heap.
If use two different variables like you suggest, after allocating buffer on the heap we will always keep in the memory buffer ( on the stack),  due to there is no options to free it. What the clue of holding garbage inside instance that will never be used?

Reply:
I think you understand my comment wrong. you have T bufer[SmallBufferSize] on stack. When you need bigger array than SmallBufferSize, you should reallocate memory. buffer = new T[newCapacity]; that's it. In your case you make it overcomplicated.